onnx/numpy_helper.py

```diff
@@ -221,6 +221,98 @@ def unpack_int4(
     return res
 
 
+def evaluate_float4e2m1_from_bits(x):
+    """Evaluate the numerical value of a single float4e2m1 element represented as uint8
+    See :ref:`onnx-detail-int4` for technical details.
+
+    Args:
+        x: a uint8 element representing a float4e2m1 (using the 4 LSB)
+
+    Returns:
+        Packed array with size `ceil(farray.size/2)` (single dimension).
+    """
+    # x is stored in 4 LSB of int
+    # assert(isinstance(x, np.uint8))
+    S = -1 if bool(x & 0x08) else 1
+    M = x & 0x01
+    E = (x & 0x06) >> 1
+    if E == 0:
+        # denormalized
+        val = S * (M / 2.0)
+    else:
+        # normalized
+        val = S * (1.0 + M / 2.0) * 2.0 ** (E - 1)
+    return val
+
+
+def unpack_float4e2m1(
+    data: np.int32 | np.ndarray,
+    dims: int | Sequence[int],
+) -> np.ndarray:
+    """Converts ndarray of float4e2m1 (as packed uint8) to f32
+    See :ref:`onnx-detail-float4` for technical details.
+
+    Args:
+        data: A numpy array, empty dimensions are allowed if dims is
+            None.
+        dims: The dimensions are used to reshape the unpacked buffer
+
+    Returns:
+        A numpy array of float32 reshaped to dims.
+    """
+    single_func = lambda x: subbyte.unpack_single_4bitx2(x, False)  # noqa: E731
+    func = np.frompyfunc(single_func, 1, 2)
+
+    res_high, res_low = func(data.ravel())
+    res = np.empty((res_high.size + res_low.size,), dtype=np.float32)
+
+    def evaluate_func(x):
+        return evaluate_float4e2m1_from_bits(x)
+
+    func2 = np.frompyfunc(evaluate_func, 1, 1)
```

Comment:
frompyfunc doesn't provide much performance benefits because it is still running the python function for each element. Prefer using np operations on the whole ndarray to implement evaluate_float4e2m1_from_bits

Reply:
That should be ok. The code is easier to read and convert into C++ in that form. The reference implementation is not meant to be the most efficient one.

Reply:
We use the reference runtime to do constant folding, and this will make it too expensive to be usable. I agree the reference runtime doesn’t need to be the most performant, but it’s good to not be much slower than native numpy. 

@yuanyao-nv I suggest keeping the single element implementation as a reference so we can use it to test, but publicly expose a version that is properly vectorized and use that in the reference implementation.

Reply:
@justinchuby I was following what was done for the other custom dtypes (fp8, int4/uint4). They are also using frompyfunc/vectorize in the implementation. Have you experienced perf issues with those as well?

Reply:
Yes. That was what I was calling out in the meeting. vectorize is numpy applying the _python_ function over all elements so that's slow. My other PR was to fix the existing code but it was too big so I am taking my time to break it down.

Reply:
I removed some instances of `frompyfunc`/`vectorize`, including this one.